### PR TITLE
[OGR provider] Revise significantly the way we handle subset filter to avoid issues with FID (fixes #20136)

### PR DESCRIPTION
**Pull request by rouault**
Some rationale on this change...

Previously when applying a "regular" subset string, ie. one that is only the
content of a where clause, we issued a full "SELECT * FROM layer WHERE subsetstring",
resulting in a OGR SQL layer. The caveat of that is that most OGR drivers
will have issues retaining the original FID. A hack consisting in adding a
{original_fid_name} as orig_ogc_fid to the select columns was introduced in
https://github.com/qgis/QGIS/commit/4ce2cf1744b008043403b18b8def8f18c99d14f1
to try to retain the original FID, but this added a lot of complexity. And
actually, in the case of the OGR GPKG driver, it caused it to still be confused
when analyzing the column definition of the resulting layer, since it sees
2 FID columns despite the renaming (one included in the '*' wildcard, and the
one of orig_ogc_fid), which caused it to use sequential FID numbering (the
driver when seeing more than once a column that is the FID column assumes that
some cross join is done, and thus that FID are unreliable)

A simpler and more robust (crossing fingers!) approach in that case is
just to use OGR_L_SetAttributeFilter() instead of GDALDatasetExecuteSQL().
Some care must be taken to cancel the filter when removing the subset
filter, or in QgsOgrFeatureIterator when combining with the filter
expression coming from the request, but besides that, this is more
straightforward, and actually solves #20136

### CODE REVIEW

**Review by nyalldawson — COMMENTED** [review]
Thanks! This was getting ridiculously complex (and more and more fragile). I can't comment on the details here, but I'm confident we have good test coverage with this so if Travis is happy than I'm +1

**Comment by nirvn** [comment]
Nice. I'll test this out and report.

**Comment by nirvn** [comment]
@rouault , spotted a regression, layer not rendered. Will provide a reduced test project.

**Comment by nirvn** [comment]
@rouault , check this test project:
[filter.zip](https://github.com/qgis/QGIS/files/2490434/filter.zip)

The shapefile layer is filtered to show one single polygon. While the feature is visible in the attribute table, it's not rendered on the canvas.

**Comment by rouault** [comment]
@nirvn Thanks for the testing and report. I've updated my commit with a fix, and a new test case in providertestbase.py that is representative of the rendering request ( I did my tests with a dummy "1 = 1" filter that didn't hit that bug... )

**Comment by rouault** [comment]
the fix for the rendering situation consists into https://github.com/qgis/QGIS/pull/8216/commits/4dd8fb90fb7ef08d379c207eb09903c24f342e7e#diff-63b6a1d9be718d66b05c5f314071c51eR1131 . Couldn't really think of something better.

**Comment by nirvn** [comment]
@rouault , building as we speak, will report soon.

**Comment by elpaso** [comment]
@rouault that reminds me of a similar issue on another provider: https://github.com/qgis/QGIS/pull/8052/files#diff-e51dad57afaacf769bd4b1e063ca7337R4593

**Comment by nirvn** [comment]
@rouault , OK, things are looking good when testing random projects on my machine. Good job! :)

**Comment by nirvn** [comment]
Exciting times, great to see the edit action not disabled for filtered shapefile et cie layers anymore.  We should make sure it's highlighted in the visual changelog.

**Comment by m-kuhn** [comment]
@rouault if you can cherry-pick 506762d343e4da3e5324a498dc646d98d6f60175 I can close the other PR (and this one is better tested)

**Comment by rouault** [comment]
Done. https://github.com/qgis/QGIS/commit/506762d343e4da3e5324a498dc646d98d6f60175 cherry-picked here

**Comment by nyalldawson** [comment]
I think we should merge asap. It's an important fix, but likely one which will only get fully stress tested after merge.

File: src/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -1132,7 +1128,17 @@ void QgsOgrProviderUtils::setRelevantFields( OGRLayerH ogrLayer, int fieldCount,
...
           const char *fieldName = OGR_Fld_GetNameRef( field );
-          if ( qstrcmp( fieldName, ORIG_OGC_FID ) != 0 )
+          // This is implemented a bit in a hacky way, but in case we are acting on a layer
+          // with a subset filter, do not ignore fields that are found in the
+          // where clause. We do this in a rough way, by looking, in a case
+          // insensitive way, if the current field name is in the subsetString,
+          // so we potentially don't ignore fields we could, in situations like
+          // subsetFilter == "foobar = 2", and there's a "foo" or "bar" field.
+          // Better be safe than sorry.
+          // We could argue that OGR_L_SetIgnoredFields() should be aware of
```
Comment (nyalldawson) [inline src/providers/ogr/qgsogrprovider.cpp]: s/could/should/

FWIW, it's what QGIS expression based feature requests do -- we silently add any attributes required for client side filtering and sorting to the request.

**Comment by DelazJ** [comment]
> Exciting times, great to see the edit action not disabled for filtered shapefile et cie layers anymore. We should make sure it's highlighted in the visual changelog.

Agreed! we'd also need to update the user manual that states that edit is not possible for some filtered formats.

**Pull request merged**